File: linked_list.py
class Node:
  '''
  an object for storing a single node of a linked list.
  Models two attributes - data and the link to the next node in the list
  '''
  data = None
  next_node = None

  def __init__(self, data):
    self.data = data

  # def __str__(self):
  #     return '\
  #     <Node data: {0}\n\
  #     \tnext_node: {1}>\n\
  #     '.format(str(self.data), self.next_node)

  def __repr__(self):
    return "<Node data: %s>" % self.data

  def getNext(self):
    return self.next_node

class LinkedList:
  '''
  singly-linked list
  '''
  def __init__(self):
    self.head = None

  def is_empty(self):
    return self.head == None

  def size(self):
    '''
    returns number of nodes in list;getting the size/length is an O(n) operation for linked-lists
    '''
    count = 0
    current = self.head
    while current:
      print(current.data)
      count += 1
      current = current.next_node

    return count 

  def add(self, data):
    '''
    prepends data to the head of the list O(1)
    '''
    current_head = self.head
    self.head = Node(data)
    self.head.next_node = current_head

  def search(self, data):
    '''
    traverses list in search of data
    returns node containing data if found otherwise None
    takes O(n) time
    '''
    current = self.head
    while current:
      if current.data == data: return current
      current = current.next_node

    return None

  def insert(self, data, position):
    '''
    inserts data at position
    position starts at 0 (0 == head)
    inserting take O(1) time but finding the correct position takes O(n) time
    '''
    new_node = Node(data)
    count = 0
    current = self.head
    previous = None


    print('position = {0}'.format(position))
    while current:
      print('count = {0}'.format(count))
      print('current = {0}'.format(current))
      if position == 0:
          #when inserting a new head
          self.add(data)
          return count
      elif count == position:        
        previous.next_node = new_node 
        previous.next_node.next_node = current
        return count
      elif current.next_node == None:
        print(current)
        print(previous)
        current.next_node = new_node
        return count
      

      count += 1
      previous = current
      current = current.next_node

    return -1

  def remove(self, data):
    '''
    Removes the node from the list that contains data
    '''
    current = self.head
    previous = None

    while current:
      if current.data == data:
        if current is self.head:
          self.head = current.next_node
        else:
          previous.next_node = current.next_node

        del current
        return

      previous = current
      current = current.next_node

  def remove_at_index(self, index): 
    '''
    removes the node at index
    '''
    current = self.head
    previous = None
    count = 0
    while current:
      print('current = {0}'.format(current))
      if count == index:
        if current is self.head:
          self.head = current.next_node
        else:
          previous.next_node = current.next_node

        del current
        return

      previous = current
      current = current.next_node
      count += 1

  def get_node_at_index(self, index):
    '''
    returns the node at index
    '''
    current = self.head
    i = 0

    while current:
      if index == i: return current
      i+=1
      current = current.next_node

  def __repr__(self):
    '''
    returns a string representation of the list 
    take 0(n) time
    '''

    nodes = []
    current = self.head
    
    while current:
      if current is self.head:
        nodes.append("[Head: %s]" % current.data)
      elif current.next_node is None:
        nodes.append("[Tail: %s]" % current.data) 
      else:
        nodes.append("[%s]" % current.data)

      current = current.next_node
    return "->".join(nodes)

  def __eq__(self, other):
    if isinstance(other, LinkedList):
      if self.__repr__() == other.__repr__(): return True
    return False

linked_list = LinkedList()

# linked_list.add(20)
# linked_list.add(22)
# linked_list.add(30)
linked_list.add(31)
linked_list.add(32)
linked_list.add(33)
linked_list.insert(34,-1)
linked_list.remove(34)
linked_list.remove_at_index(0)
print(linked_list)
print(linked_list.get_node_at_index(1))

File: LinkedList.ts
//My implementation of a LinkedList using TypeScript
import { InstanceTypes } from './Stack.js'
class LinkedListNode <T> {
  public data: T;
  public nextNode: any = null;
  public previousNode: any = null;

  constructor(data: T) {
    this.data = data;
  }
}
class LinkedList <T> {
  private _TYPE_ERROR: string = 'Incorect Type!';
  private _type: InstanceTypes;
  private _head: LinkedListNode <T>;
  private _tail: LinkedListNode <T>;
  private _size: number = 0;

  // public tail: LinkedListNode<T>;
  
  get size() {
    return this._size;
  }
  get type() {
    return this._type;
  }
  get head() {
    return this._head;
  }
  get tail() {
    return this._tail;
  }

  constructor(type: InstanceTypes) {
    this._type = type;
    this._head = new LinkedListNode(null) as any;
    this._tail = new LinkedListNode(null) as any;
  }

  private _insert(data: T, nodeToInsertAt: LinkedListNode <T>, isInsertAtHead: boolean = true) {
    //Inserts at head O(1)
    if (typeof data !== this._type) return this._TYPE_ERROR;
    const newHead = new LinkedListNode(data);

    if (nodeToInsertAt === this._tail) {
      newHead.previousNode = nodeToInsertAt;
      nodeToInsertAt.previousNode = newHead;
      this._tail = newHead;
    }
    else {
      newHead.nextNode = nodeToInsertAt;
      const tempPrevious = nodeToInsertAt.previousNode;
      nodeToInsertAt.previousNode = newHead;
      newHead.previousNode = tempPrevious;

      if (isInsertAtHead) this._head = newHead;
      else tempPrevious.nextNode = newHead;
    }

    if (this._size === 1) {
      this._tail.previousNode = newHead;
    }

    this._size++;
  }

  public insert(data: T){
    if (this._head.data == null) {
      this._size++;
      this._tail.data = data;
      return this._head.data = data;
    }
    
    this._insert(data, this._head);
  }

  public insertAtIndex(data: T, index: number) {
    //inserts at index if it exists otherwise, inserts after tail
    
    if (this._head.data == null) {
      this._size++; 
      this._tail.data = data;
      return this._head.data = data;
    }

    if (index > this._size - 1) {
      return this._insert(data, this._tail);
    }

    let traverser = this._head;
    let count = 0
    while (count !== index) {
      traverser = traverser.nextNode;
      count++;
    }

    this._insert(data, traverser, false);
  }

  public searchFromStart(data: T) {
    if (typeof data !== this._type) return this._TYPE_ERROR;
    
    let traverser = this._head;
    let count = 0
    while (traverser) {
      if (data === traverser.data) return count;
      count++;
      traverser = traverser.nextNode;
    }
    return -1;
  }

  public searchFromEnd(data: T) {
    if (typeof data !== this._type) return this._TYPE_ERROR;
    
    let traverser = this._tail;
    let count = 0
    while (traverser.data) {
      if (data === traverser.data) return this._size - count;
      count++;
      traverser = traverser.previousNode;
    }
    return -1;
  }

  public delete() {
    this._size--;
  }

  public deleteAt() {
    this._size--;
  }

  public viewList(startFromHead: boolean =  true) {
    const values = []

    if (startFromHead) {
      let start = this._head;
      while(start){
        values.push(start.data)
        start = start.nextNode
      }
    }
    else {
      let start = this._tail;
      while(start){
        values.push(start.data)
        start = start.previousNode
      }
    }
    
    console.log(values.join('=>'))
  }

}

const ll = new LinkedList(InstanceTypes.number);
ll.insert(1)
ll.insert(2)
ll.insert(3)
ll.insert(4)
ll.insertAtIndex(31,1)
ll.insertAtIndex(32,1)
ll.insertAtIndex(33,4)

// ll.insertAtIndex(40,3)
// ll.insertAtIndex(5,4)
// ll.insertAtIndex(6,4)
// ll.insert(2)
// ll.insert(3)
ll.viewList();
ll.viewList(false);
// console.log('ll.size =', ll.size);

// ll.insert(3)
// ll.insert(4)